### PR TITLE
rewrite using ssb-feed-format

### PR DESCRIPTION
**Pull request by staltz**
Part of https://github.com/ssbc/ssb-db2/pull/347

@arj03 I'm curious how performance will be with this implementation. I couldn't test the perf myself because the benchmarks in ssb-db2 don't use add() (which would trigger addBatch()), so I think you'd have to test this in ssb-ebt with that "3s" test you have.

### CODE REVIEW

**Comment by arj03** [comment]
Will get to you on this tonight or tomorrow

File: format.js [hunk 1]
```diff
@@ -0,0 +1,327 @@
...
+  }
+  const [encodedValue] = extract(nativeMsg)
+  const [authorBFE, parentBFE] = extractVal(encodedValue)
+  const author = bfe.decode(authorBFE)
+  const parent = bfe.decode(parentBFE)
+  if (parent) {
+    const { data } = SSBURI.decompose(parent)
+    const feedId = author + '/' + data
+    _feedIdCache.set(nativeMsg, feedId)
+    return author + '/' + data
```
Comment (arj03) [inline format.js]: return feedId

File: format.js [hunk 2]
```diff
@@ -0,0 +1,327 @@
...
+    sequence,
+    timestamp,
+    previousBFE,
+    tag,
+    contentBuffer.length,
+    contentHash,
+  ]
+
+  const encodedValue = bipf.allocAndEncode(value)
+  // FIXME: we need ssb-keys to support returning buffer from sign()
```
Comment (arj03) [inline format.js:116]: Yes! Good idea

Comment (staltz) [reply format.js:116]: The only reason why I didn't do it is because https://github.com/ssbc/ssb-bfe-spec/issues/24 :smile:

File: format.js [hunk 3]
```diff
@@ -0,0 +1,327 @@
...
+    tag,
+  }
+  const bipfMsg = bipf.allocAndEncode(msgVal)
+  _bipfMsgValCache.set(nativeMsg, bipfMsg)
+  return bipfMsg
+}
+
+function fromNativeMsg(nativeMsg, encoding = 'js') {
+  if (encoding === 'js') {
+    return _fromNativeMsgJS(nativeMsg)
```
Comment (arj03) [inline format.js]: Could we call this `_fromNativeToMsgJS` instead. Same with `_fromNativeToMsgBIPF` ?

Comment (staltz) [reply format.js]: Done

File: format.js [hunk 4]
```diff
@@ -0,0 +1,327 @@
...
+    contentLen,
+    contentHash,
+  ]
+  const encodedValue = bipf.allocAndEncode(value)
+  return bipf.allocAndEncode([encodedValue, sigBuf, contentBuffer])
+}
+
+function toNativeMsg(msgVal, encoding = 'js') {
+  if (encoding === 'js') {
+    return _toNativeMsgJS(msgVal)
```
Comment (arj03) [inline format.js]: Similar here, `_toNativeFromMsgJS` and `_toNativeFromMsgBIPF`

Comment (staltz) [reply format.js]: already done :sunglasses:

File: package.json [hunk 5]
```diff
@@ -7,22 +7,44 @@
...
+    "!example.js"
+  ],
   "dependencies": {
     "bipf": "^1.7.0",
     "blake3": "^2.1.7",
     "fast-varint": "^1.0.1",
     "monotonic-timestamp": "^0.0.9",
     "ssb-bfe": "^3.3.0",
-    "ssb-keys": "^8.3.0"
+    "ssb-keys": "^8.3.0",
```
Comment (arj03) [inline package.json]: can we bump this to 8.4? tests are not running with 8.3

Comment (staltz) [reply package.json]: Done

File: test/basic.js [hunk 6]
```diff
@@ -1,192 +1,238 @@
...
-    const prevData = i === 0 ? null : datas[i-1]
-    const prevMsgKey = i === 0 ? null : msgKeys[i-1]
-    const keys = butt2.validateBatch(datas.slice(i, i+25), prevData, prevMsgKey, hmacKey)
-    for (var j = 0; j < keys.length; ++j)
-      result.push(keys[j])
+  for (let i = 0; i < N; i += 100) {
+    const prevNativeMsg = i === 0 ? null : nativeMsgs[i - 1]
+    if (
+      (err = butt2.validateBatchSync(
+        nativeMsgs.slice(i, i + 25),
```
Comment (arj03) [inline test/basic.js]: this should be i + 100, right?

Comment (staltz) [reply test/basic.js]: Ouch, good point. I fixed it now, but batch is (almost always) slower than single :(

Comment (staltz) [reply test/basic.js]: I fixed this, but the tests are probably going to fail often.

Comment (staltz) [reply test/basic.js]: Ah, nevermind! I think I found the perf problem.

Comment (staltz) [reply test/basic.js]: Niiice

```
validateSingle completes in 319 ms
validateBatch completes in 56 ms
```

File: validation.js [hunk 7]
```diff
@@ -0,0 +1,296 @@
...
+    return new Error('invalid hmac')
+  }
+
+  if (bytes.length !== 32) {
+    return new Error('invalid hmac, it should have 32 bytes')
+  }
+}
+
+function _validateSize(nativeMsg) {
+  if (nativeMsg.length > 8192) {
```
Comment (arj03) [inline validation.js]: The spec says 16384. I once asked Dominic about the size and he said, no matter what number you pick they will always ask for more, but I was thinking with something like crut that does a lot of tangle stuff, the 8k can be a bit limiting, so why now bump it just a bit.

Comment (staltz) [reply validation.js]: Fixed

**Comment by arj03** [comment]
I'm 95% sure this is good. I'll have to do a little more testing and then I'll do the ebt test before pressing the big green button. Time for dinner now :)

**Comment by arj03** [comment]
Pushed up a test because I wanted to make sure that parent works correctly with the feed format messages

**Review by arj03 — APPROVED** [review]
Thanks, this looks good. I'm quite happy with the way ssb-feed-formats have turned out. I have plugged this into the ebt tests and things are running fine. I'll report numbers in that branch

**Comment by staltz** [comment]
Thanks for reviewing! Lots of changes going on lately. But this should be one of the last ones.

**Comment by staltz** [comment]
Can you publish ssb-buttwoo@0.3.0 or give me publish rights?

**Comment by arj03** [comment]
@staltz you have been invited :)

**Pull request merged**